python/engine/config.py

```
@@ -152,6 +165,33 @@ def _validate(self):
             raise ValueError("self_generated_values requires projection_kind='adaptive'")
         if self.self_ref_chunk_size > 1 and self.projection_kind != "adaptive":
             raise ValueError("self_ref_chunk_size > 1 requires projection_kind='adaptive'")
+        # V-01: retention–rule compatibility
+        if self.retention == "sphere_normalization" and self.memory_rule != "lattice":
+            raise ValueError(
+                f"retention 'sphere_normalization' requires memory_rule 'lattice_osr', "
+                f"got '{self.memory_rule}'. Use retention 'l2_weight_decay' for {self.memory_rule}.")
```

Ruff (0.15.2)

[warning] Avoid specifying long messages outside the exception class

(TRY003)